daemon/cmd/config.go

```diff
@@ -87,7 +87,7 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 	if changes > 0 {
 		// Only recompile if configuration has changed.
 		log.Debug("daemon configuration has changed; recompiling base programs")
-		if err := d.Datapath().Loader().Reinitialize(d.ctx, d, d.mtuConfig.GetDeviceMTU(), d.Datapath(), d.l7Proxy); err != nil {
+		if err := d.Datapath().Loader().Reinitialize(d.ctx, d, d.mtuConfig.GetRouteMTU(), d.Datapath(), d.l7Proxy); err != nil {
```

Comment:
> Somehow the per route MTU for cilium_host is not taking effect and node
port traffic is being fragmented at outer encap layer (in mix mode with
kubeproxy).

I think it'd be prudent to investigate and define how this happens exactly before making changes in the code. MTU can be tricky enough as it is, and we've arrived at the current configurations based on deployment experience across several releases. I would feel much more confident if we can prove we have:
1. Understanding of why this occurs, and
2. Tests that validate these changes do not cause a regression. These don't have to be new tests if we feel there is already sufficient coverage but I'm not sure where we're at with MTU scenario coverage at the moment for both cluster-internal and also external handling.

Reply:
> > Please let me know if it doesn't make sense.
> 
> You asked for it ;-)
> 
> Some background here is that Linux applies the MTU for devices on both receive and transmit sides. To account for this, we typically configure route MTUs for the transmit side to ensure that packets sourced within the cluster do not exceed the tunnel MTU when sent to other nodes. Then we configure the device MTUs higher for the receive side, so that if the node has direct access to the outside with a path MTU that is higher than the tunnel MTU, the local endpoints can still receive larger packets directly from the outside, even if they wouldn't transmit packets that are that big.

Thanks for the explanation. I see what you mean. Though I don't understand how the MTU of cilium_host matters here in tunneling mode. If cilium_host has the same MTU of the physical nic (1500), I think it only can receive packets after decap, which is of 1450 bytes. How can cilium_host receive packets bigger than that?

Reply:
> > Somehow the per route MTU for cilium_host is not taking effect and node
> > port traffic is being fragmented at outer encap layer (in mix mode with
> > kubeproxy).
> 
> I think it'd be prudent to investigate and define how this happens exactly before making changes in the code. MTU can be tricky enough as it is, and we've arrived at the current configurations based on deployment experience across several releases. I would feel much more confident if we can prove we have:
> 
> 1. Understanding of why this occurs, and
> 2. Tests that validate these changes do not cause a regression. These don't have to be new tests if we feel there is already sufficient coverage but I'm not sure where we're at with MTU scenario coverage at the moment for both cluster-internal and also external handling.

Yes. Let me dig into this. I don't have much kernel knowledge so any hints would be appreicated.

Reply:
I consulted our kernel team and confirmed this is a expected behavior. I.e. MTU on routes are not enforced for forwarding traffic. And it's just fixed last month ([patch](http://patchwork.ozlabs.org/project/netdev/patch/20200923201815.388347-1-zenczykowski@gmail.com/))

From that, I think it's the right way to decrease MTU on the device as this PR does. Let me know how we should proceed.

Reply:
I'm confused, the PR title says:

> the per route MTU for cilium_host is not taking effect 

Then in the above reply, the linked patch is named:

> always honour route mtu during forwarding

This sounds like if you apply the linked patch, then the issue is fixed.

Can you confirm which kernel version you were testing when you observed the original issue, and whether it contains the patch linked above?

Reply:
Note that the patch is only accepted on 23 Sep, 2020. It's not widely available yet.

Reply:
That patch is queued up for -stable so it will propagate to all supported kernels over the coming days. Can you confirm whether it fixes your issue without additional changes to Cilium?

Reply:
So I patched kernel v5.4.61 manually and verified that it fixes the issue.

Even if the patch gets accepted by stable and included in new kernel revisions for all ongoing kernel series, I don't see it will be present widely soon in user environments and cilium shouldn't rely on that.

Reply:
Thanks for confirming the fix, that's great news.

> Even if the patch gets accepted by stable and included in new kernel revisions for all ongoing kernel series, I don't see it will be present widely soon in user environments and cilium shouldn't rely on that.

A kernel bug was identified, and you've pointed out the patch that fixes the bug. If it's a kernel bug, then a kernel fix is necessary. If someone hits this issue with MTU with Cilium, then we can point to that kernel patch. If that patch is not available in a user's distribution, then the user should follow up with their Linux distribution vendor to ensure that the patch is backported to a version they can use. You are of course welcome to proactively follow up with distros in advance to speed this process along and reduce the likelihood that other users will hit this bug.

Cilium's general approach to MTU is matching the expectations outlined by the core kernel folks in the patch thread linked above, ie configure device MTU for rx, route MTU for tx. I don't think it's a good idea to follow through with this PR to violate those base assumptions, it is likely to just move the problem elsewhere.

Reply:
Given that, I think it's worthwhile to make this visible in cilium docs like [here](https://docs.cilium.io/en/v1.8/concepts/networking/fragmentation/) so that others don't waste time debugging this issue.

Even the bug is in kernel, this is still an unacceptable gap for us. We will move ahead with internal patches. If you see anything that is broken with this, please let us know.

Reply:
Sounds good, and I like your idea about mentioning it in the docs. Would you mind submitting the change for that?